drivers/sensor/sensirion/sen6x/sen6x_decoder.c

```diff
@@ -0,0 +1,461 @@
+		.has_measured_values = needs_measured_values,
+		.has_number_concentration = needs_separate_number_concentration,
+	};
+
+	return 0;
+}
+
+static int sen6x_decoder_get_frame_count(const uint8_t *buffer, struct sensor_chan_spec chan_spec,
+					 uint16_t *frame_count)
+{
+	struct sen6x_encoded_data *edata = (struct sen6x_encoded_data *)buffer;
+	int ret;
+	uint8_t *channel_data;
+	uint16_t value;
+	int16_t value_signed;
+	bool is_signed;
+
+	if (chan_spec.chan_idx != 0) {
+		return -ENOTSUP;
+	}
+
+	ret = sen6x_get_channel_index(edata, chan_spec.chan_type);
+	if (ret < 0) {
+		return ret;
+	}
+	if (ret < MAX_MEASURED_VALUES_COUNT && !edata->header.has_measured_values) {
+		return -ENODATA;
+	}
+	if (ret >= MAX_MEASURED_VALUES_COUNT && !edata->header.has_number_concentration) {
+		return -ENODATA;
+	}
+
+	ret = sen6x_get_channel_format(chan_spec.chan_type, &is_signed, NULL);
+	if (ret) {
+		return ret;
+	}
+	channel_data = &edata->channels[ret * 3];
+
+	value = sys_get_be16(channel_data);
+	value_signed = (int16_t)value;
+
+	if (is_signed) {
+		if (value == INT16_MAX) {
+			return -ENODATA;
+		}
+	} else {
+		if (value == UINT16_MAX) {
+			return -ENODATA;
+		}
+	}
+
+	*frame_count = 1;
+	return 0;
+}
+
+static int sen6x_decoder_get_size_info(struct sensor_chan_spec chan_spec, size_t *base_size,
+				       size_t *frame_size)
+{
+	switch (chan_spec.chan_type) {
+	case SENSOR_CHAN_PM_1_0:
+	case SENSOR_CHAN_PM_2_5:
+	case SENSOR_CHAN_PM_4_0:
+	case SENSOR_CHAN_PM_10:
+	case SENSOR_CHAN_PM_0_5_NUM:
+	case SENSOR_CHAN_PM_1_0_NUM:
+	case SENSOR_CHAN_PM_2_5_NUM:
+	case SENSOR_CHAN_PM_4_0_NUM:
+	case SENSOR_CHAN_PM_10_NUM:
+	case SENSOR_CHAN_AMBIENT_HUMIDITY:
+	case SENSOR_CHAN_AMBIENT_TEMP:
+	case SENSOR_CHAN_VOC_INDEX:
+	case SENSOR_CHAN_NOX_INDEX:
+	case SENSOR_CHAN_CO2:
+	case SENSOR_CHAN_HCHO:
+		*base_size = sizeof(struct sensor_q31_data);
+		*frame_size = sizeof(struct sensor_q31_sample_data);
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int sen6x_one_shot_decode(const uint8_t *buffer, struct sensor_chan_spec chan_spec,
+				 uint32_t *fit, uint16_t max_count, void *data_out_)
+{
+	struct sen6x_encoded_data *edata = (struct sen6x_encoded_data *)buffer;
+	struct sensor_q31_data *data_out = data_out_;
+	int ret;
+	uint8_t *channel_data;
+	uint16_t value;
+	int16_t value_signed;
+	size_t divisor;
+	bool is_signed;
+
+	if (*fit != 0) {
+		return 0;
+	}
+
+	if (max_count == 0 || chan_spec.chan_idx != 0) {
+		return -EINVAL;
+	}
+
+	ret = sen6x_get_channel_index(edata, chan_spec.chan_type);
+	if (ret < 0) {
+		return ret;
+	}
+	if (ret < MAX_MEASURED_VALUES_COUNT && !edata->header.has_measured_values) {
+		return -ENODATA;
+	}
+	if (ret >= MAX_MEASURED_VALUES_COUNT && !edata->header.has_number_concentration) {
+		return -ENODATA;
+	}
+
+	channel_data = &edata->channels[ret * 3];
+
+	ret = sen6x_get_channel_format(chan_spec.chan_type, &is_signed, &divisor);
+	if (ret) {
```

Comment:
Please avoid implicit boolean conversions, [rule](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#misrac-rule-14-4)

```suggestion
	if (ret < 0) {
```